src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs

```diff
@@ -174,84 +204,107 @@ private BoundSequence VisitIndexOrRangePatternIndexerAccess(BoundIndexOrRangePat
-            BoundExpression argument,
-            bool isLeftOfAssignment)
+        private BoundSequence VisitIndexImplicitIndexerAccess(BoundIndexOrRangePatternIndexerAccess node, bool isLeftOfAssignment)
         {
+            Debug.Assert(node.ArgumentPlaceholders.Length == 1);
+            Debug.Assert(node.IndexerAccess is BoundIndexerAccess);
+
+            Debug.Assert(TypeSymbol.Equals(
+                node.Argument.Type,
+                _compilation.GetWellKnownType(WellKnownType.System_Index),
+                TypeCompareKind.ConsiderEverything));
+
             var F = _factory;
+            var receiver = node.Receiver;
 
             Debug.Assert(receiver.Type is { });
             var receiverLocal = F.StoreToTemp(
                 VisitExpression(receiver),
                 out var receiverStore,
                 // Store the receiver as a ref local if it's a value type to ensure side effects are propagated
                 receiver.Type.IsReferenceType ? RefKind.None : RefKind.Ref);
-            var indexAccess = MakePatternIndexOffsetExpression(argument, F.Property(receiverLocal, lengthOrCountProperty), out _);
 
-            return (BoundSequence)F.Sequence(
-                ImmutableArray.Create<LocalSymbol>(receiverLocal.LocalSymbol),
-                ImmutableArray.Create<BoundExpression>(receiverStore),
-                MakeIndexerAccess(
-                    syntax,
-                    receiverLocal,
-                    intIndexer,
-                    ImmutableArray.Create<BoundExpression>(indexAccess),
-                    default,
-                    default,
-                    expanded: false,
-                    argsToParamsOpt: default,
-                    defaultArguments: default,
-                    intIndexer.Type,
-                    oldNodeOpt: null,
-                    isLeftOfAssignment));
+            var receiverPlaceholder = node.ReceiverPlaceholder;
+            AddPlaceholderReplacement(receiverPlaceholder, receiverLocal);
+
+            Debug.Assert(node.LengthOrCountAccess is not null);
+            var integerArgument = MakePatternIndexOffsetExpression(node.Argument, VisitExpression(node.LengthOrCountAccess), out _);
+            Debug.Assert(integerArgument.Type!.SpecialType == SpecialType.System_Int32);
+
+            if (isLeftOfAssignment && node.IndexerAccess.GetRefKind() == RefKind.None)
+            {
+                RemovePlaceholderReplacement(receiverPlaceholder);
+
+                // We produce an unlowered indexer access which be rewritten 
+                // with MakePropertyAssignment when rewriting the enclosing BoundAssignmentOperator.
+                var indexerAccess = (BoundIndexerAccess)node.IndexerAccess;
+                return (BoundSequence)F.Sequence(
+                    ImmutableArray.Create<LocalSymbol>(receiverLocal.LocalSymbol),
+                    ImmutableArray.Create<BoundExpression>(receiverStore),
+                    MakeIndexerAccess(
```

Comment:
>MakeIndexerAccess [](http://example.com/codeflow?start=20&length=17)

This doesn't look right. I think we shouldn't be using this helper here. instead, we should always rewrite ```node.IndexerAccess``` without any conditional logic related to ```isLeftOfAssignment```. Since we can be certain that ```node.IndexerAccess``` is always a ```BoundIndexerAccess``` here, we should call ```VisitIndexerAccess``` directly and pass ```isLeftOfAssignment``` to it.